src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+                ." FROM {$this->tableItems} item LEFT JOIN 
+{$this->tableVersions} version ON item.id = version.id AND item.version = version.version AND item.sequence > :sequence"
+                .' ORDER BY item.sequence LIMIT 100',
+                ['sequence' => $cursor ?? 0]
+            );
+
+            while ($result = $results->fetch(FetchMode::ASSOCIATIVE)) {
+                yield $this->rowToItemVersion($result);
+                $cursor = $result['sequence'];
+            }
+
+            if (empty($result)) {
+                break;
+            }
+        }
+    }
+
+    public function getSchema() : Schema
+    {
+        $schema = new Schema();
+
+        $table = $schema->createTable($this->tableItems);
+        $table->addColumn('id', Type::STRING);
+        $table->addColumn('version', Type::INTEGER);
+        $table->addColumn('sequence', Type::INTEGER, ['autoincrement' => true]);
+        $table->setPrimaryKey(['sequence']);
+        $table->addUniqueIndex(['id']);
+
+        $table = $schema->createTable($this->tableVersions);
+        $table->addColumn('id', Type::STRING);
+        $table->addColumn('version', Type::INTEGER);
+        $table->addColumn('content', Type::BLOB);
+        $table->addColumn('hash', Type::STRING);
+        $table->setPrimaryKey(['id', 'version']);
+        $table->addIndex(['id']);
+
+        return $schema;
+    }
+
+    private function rowToItemVersion(array $row) : ItemVersion
+    {
+        ['id' => $id, 'version' => $version, 'content' => $content, 'hash' => $hash] = $row;
+
+        // Some drivers, eg pdo_sqlite, unfortunately return strings rather than resources.
+        // Doctrine doesn't provide an equivalent of `PDOStatement::bindColumn()` either.
+        // This might cause memory problems if the content is too large.
+        $content = Type::getType('blob')->convertToPHPValue($content, $this->connection->getDatabasePlatform());
+
+        return new ItemVersion(ItemId::fromString($id), ItemVersionNumber::create($version), $content, $hash);
+    }
+
+    private function getCurrentVersion(ItemId $id, int $lockMode = LockMode::NONE) : ItemVersionNumber
+    {
+        $platform = $this->connection->getDatabasePlatform();
+
+        $sql = 'SELECT item.version FROM %s WHERE item.id = :id %s';
+        if (LockMode::NONE === $lockMode) {
+            $sql = sprintf($sql, "{$this->tableItems} item", '');
+        } else {
+            $sql = sprintf(
+                $sql,
+                $platform->appendLockHint("{$this->tableItems} item", LockMode::PESSIMISTIC_READ),
+                $platform->getWriteLockSQL()
+            );
+        }
+
+        $stmt = $this->connection->executeQuery($sql, ['id' => (string) $id]);
+        $current = $stmt->fetch(FetchMode::COLUMN);
+
+        if (false === $current) {
+            throw new ItemNotFound($id);
+        }
+
+        return ItemVersionNumber::create($current);
+    }
+
+    private function deleteItem(ItemId $id) : void
```

Comment:
a bit dangerous that the two private `delete*()` assume an open transaction

Reply:
Could wrap all these in transactions. As mentioned in https://github.com/libero/content-api-bundle/pull/11#discussion_r231027022 they won't be real DB transactions.